### ISSUE TITLE
`machinectl set-limit infinity` fails with unclear message on raw-backed machine volume

### ISSUE THREAD

**Issue by eliasp** [systemd/systemd#2253]
Trying to set up a new raw-backed machine volume using `machinectl set-limit infinity` (a fixed size like `6G` works just fine) fails with an unclear error message which could be made more user friendly:

```
# machinectl set-limit infinity
Could not set limit: Failed to adjust loopback limit: Invalid argument
```

Running:
- systemd 228 from git/master 5809f340fd7e5e6c76e229059c50d92e1f57e8d8
- Kernel 4.3.0

**Comment by karelzak** [comment]
set_limit() function translates "infinity" to (uint64_t) -1 and it's used by btrfs_resize_loopback_fd() for ftruncate().

The problem is that kernel loff_t is _signed_ 64bit number, and for length < 0 kernel returns returns EINVAL. This is well known VFS limitation. example:

```
INT64_MAX:
    ftruncate(3, 9223372036854775807)       = 0
INT64_MAX + 1ULL:
    ftruncate(3, 9223372036854775808)       = -1 EINVAL (Invalid argument)
```

We need something like:

```
diff --git a/src/machine/machinectl.c b/src/machine/machinectl.c
index 685bbaf..819907f 100644
--- a/src/machine/machinectl.c
+++ b/src/machine/machinectl.c
@@ -2312,7 +2312,12 @@ static int set_limit(int argc, char *argv[], void *userdata) {
         int r;

         if (STR_IN_SET(argv[argc-1], "-", "none", "infinity"))
-                limit = (uint64_t) -1;
+                /*
+                 * Linux kernel loff_t is *signed* 64bit number, and for length < 0
+                 * kernel returns EINVAL. This is VFS limitation (8EiB),
+                 * for example btrfs internal limit is 16 EiB.
+                 */
+                limit = INT64_MAX;
         else {
                 r = parse_size(argv[argc-1], 1024, &limit);
                 if (r < 0)
```

... but frankly I have doubts that "infinity" translated to any hardcoded number makes any sense. The limit still depends on device size etc. 

Maybe it would be better to improve btrfs_resize_loopback_fd() and ask for a real available space on the filesystem. I have doubts people have 8EiB (INT64_MAX) devices...

**Comment by poettering** [comment]
A fix for this is waiting in #2445

**Comment by poettering** [comment]
#2445 has been merged. CLosing
